onnxruntime/contrib_ops/cpu/signal/window_functions.cc

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef BUILD_MS_EXPERIMENTAL_OPS
+
+#include "core/providers/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/util/math_cpuonly.h"
+#include "Eigen/src/Core/Map.h"
+#include "window_functions.h"
+#include <functional>
+
+#include "core/platform/threadpool.h"
+
+#include <complex>
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    HannWindow,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    HannWindow);
+
+ONNX_OPERATOR_KERNEL_EX(
+    HammingWindow,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    HammingWindow);
+
+ONNX_OPERATOR_KERNEL_EX(
+    BlackmanWindow,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    BlackmanWindow);
+
+
+ONNX_OPERATOR_KERNEL_EX(
+    MelWeightMatrix,
+    kMSExperimentalDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().MayInplace(0, 0)
+        .TypeConstraint("T1", BuildKernelDefConstraints<int64_t>())
+        .TypeConstraint("T2", BuildKernelDefConstraints<float>())
+        .TypeConstraint("T3", BuildKernelDefConstraints<float, double, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t>()),
+    MelWeightMatrix);
+
+
+template <typename T>
+static Status cosine_sum_window(Tensor* Y, size_t size, float a0, float a1, float a2) {
+  auto* Y_data = reinterpret_cast<T*>(Y->MutableDataRaw());
+
+  // Calculate the radians to increment per sample
+  constexpr double pi = 3.14159265;
```

Comment:
is this not defined somewhere else already? also, how did you choose this level of precision?

Reply:
This was copied from the constants in DX.